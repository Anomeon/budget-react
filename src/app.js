import React from 'react';
import ReactDOM from 'react-dom';
import {
  Router,
  Route,
  hashHistory
} from 'react-router';
import {
  EnterData,
  Menu,
  Report,
  Categories
} from './components';

ReactDOM.render(
  <Router history={hashHistory}>
    <Route component={Menu}>
      <Route path="/" component={EnterData}/>
      <Route path="/items/:category" component={EnterData}/>
      <Route path="test" component={Report}/>
      <Route path="categories" component={Categories}/>
    </Route>
  </Router>,
  document.getElementById('app')
);
