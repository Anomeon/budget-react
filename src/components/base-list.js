import React from 'react';

export class BaseList extends React.Component {
  render() {
    return (
      <ul>
        {
          this.props.items.map((item) => {
            return <li key={item.id}>{item[this.props.type]}</li>
          })
        }
      </ul>
    );
  }
};
