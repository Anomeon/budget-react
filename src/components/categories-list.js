import React from 'react';

export class CategoriesList extends React.Component {
  render() {
    return (
      <ul>
        {
          this.props.items.map((item) => {
            return <li key={item.id}>{item.category}</li>
          })
        }
      </ul>
    );
  }
};
