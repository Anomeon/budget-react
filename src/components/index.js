export {EnterData} from './enter-data';
export {ListItem} from './list-item';
export {List} from './list';
export {Menu} from './menu';
export {Test} from './test';
export {Categories} from './categories';
export {CategoriesList} from './categories-list';
