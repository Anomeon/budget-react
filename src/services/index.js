export {ItemStorage} from './item-storage'
export {Piechart} from './piechart'
