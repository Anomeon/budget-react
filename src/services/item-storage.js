export class ItemStorage {
  // static setStaticProperties() {
  //   ItemStorage.instanceCounter = ItemStorage.instanceCounter || 0;
  // }

  constructor(storage) {
    if (storage) {
      this.storage = storage;
    } else {
      throw new Error('ItemStorage constructor expected a storage parameter');
    }
    // ItemStorage.setStaticProperties();
    //
    // if (ItemStorage.instanceCounter === 1) {
    //   throw new Error('ItemStorage already has instance');
    // } else {
    //   if (storage) {
    //     ++ItemStorage.instanceCounter;
    //     this.storage = storage;
    //   } else {
    //     throw new Error('ItemStorage constructor expected a storage parameter');
    //   }
    // }

    if (!this.getIDCounter()) {
      this.storage.setItem('IDCounter', 0);
    }
  }

  getItems(table, isArray, state) {
    let items = JSON.parse(this.storage.getItem(table) || '{}');
    let collection;
    if (isArray) {
      collection = [];
      for (let key in items) {
        let item = items[key];
        if (item.state === state) {
          item.id = key;
          collection.push(item);
        }
      }
    } else {
      collection = {};
      for (let key in items) {
        if (items[key].state === state) { collection[key] = items[key]; }
      }
    }
    return collection;
  }

  deleteItems(itemsArray) {
    let items = this.getItems();
    itemsArray.forEach((id) => {
      delete items[id];
    });
    this.storage.setItem('items', JSON.stringify(items));
  }

  addItem(table, itemObject) {
    let IDCounter = this.incrementIDCounter();
    let items = this.getItems(table);
    items[IDCounter] = itemObject;
    this.storage.setItem(table, JSON.stringify(items));
    return {[IDCounter]: itemObject};
  }

  updateItemState(id, state) {
    let items = this.getItems();
    items[id].state = state;
    this.storage.setItem('items', JSON.stringify(items));
  }

  getIDCounter() {
    return this.storage.getItem('IDCounter');
  }

  incrementIDCounter() {
    let IDCounter = this.getIDCounter();
    this.storage.setItem('IDCounter', ++IDCounter);
    return IDCounter;
  }
}
