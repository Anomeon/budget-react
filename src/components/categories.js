import React from 'react';
import {BaseList} from './index';
import {ItemStorage} from '../services';

export class Categories extends React.Component {
  constructor() {
    super();
    this.storage = new ItemStorage(localStorage);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.state = {
      items: this.storage.getItems('categories', true)
    }
  }

  handleSubmit(e) {
    e.preventDefault();
    let formData = {};
    [...document.querySelectorAll('input')]
      .forEach(el => formData[el.name] = el.value);
    this.storage.addItem('categories', formData);
    e.target.reset();
    this.setState({items: this.storage.getItems('categories', true)});
  }

  render() {
    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <input type="text" name="category" placeholder="Enter category" autoFocus={true}/>
          <button>Add category</button>
        </form>
        <BaseList items={this.state.items} type="category"/>
      </div>
    );
  }
};
