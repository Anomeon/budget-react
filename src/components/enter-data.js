import React from 'react';
import {ItemStorage} from '../services/item-storage';
import {List} from './index';

let storage = new ItemStorage(localStorage);

export class EnterData extends React.Component {

  constructor(props) {
    super(props);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleClearAll = this.handleClearAll.bind(this);
    this.state = {
      items: storage.getItems('items', true)
    }
  }

  handleSubmit(e) {
    e.preventDefault();
    let formData = {};
    Array.prototype.slice.call(document.querySelectorAll('input'))
      .forEach(el => formData[el.name] = el.value);
    storage.addItem('items', formData);
    e.target.reset();
    this.setState({items: storage.getItems('items', true)})
  }

  handleClearAll() {
    storage.deleteItems(
      this.state.items.map((item) => {
        return item.id;
      })
    );
    this.setState({items: storage.getItems('items', true)})
  }

  render() {
    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <label htmlFor="item"> Item </label>
          <input name="item" type="string" id="item"/>
          <label htmlFor="sum"> Sum </label>
          <input name="sum" type="number" id= "sum"/>
          <button>Send</button>
          <button type="button" onClick={this.handleClearAll}>Clear All</button>
        </form>
        <List items={this.state.items}/>
      </div>
    );
  }
};
