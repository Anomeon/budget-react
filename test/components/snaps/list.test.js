import React from 'react';
import renderer from 'react-test-renderer';
import {List} from '../../../src/components';
import {ItemStorage} from '../../../src/services/item-storage';

let storage = new ItemStorage(localStorage);

it('renders correctly', () => {
  const tree = renderer.create(
    <List items={storage.getItems('items', true)}/>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
