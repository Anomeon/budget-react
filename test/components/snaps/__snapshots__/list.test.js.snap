exports[`test renders correctly 1`] = `
<div>
  <br />
  <form>
    <input
      onChange={[Function]}
      placeholder="Type to search.." />
  </form>
  <ul>
    <li>
      <div>
        <div>
          item: 
          test
        </div>
        <div>
          sum: 
          50
        </div>
        <div>
          category: 
          test
        </div>
      </div>
    </li>
  </ul>
</div>
`;
